lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,150 @@
+
+      SUMMARY = 'Revoke a given certificate'
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca revoke [--help|--version]
+  puppetserver ca revoke [--config] --certname CERTNAME[,ADDLCERTNAME]
+
+Description:
+Given one or more valid certnames, instructs the CA to revoke them over
+HTTPS using the local agent's PKI
+
+Options:
+BANNER
+
+      def self.parser(parsed = {})
+        parsed['certnames'] = []
+        OptionParser.new do |o|
+          o.banner = BANNER
+          o.on('--certname foo,bar', Array,
+               'One or more comma separated certnames') do |certs|
+            parsed['certnames'] += certs
+          end
+          o.on('--config PUPPET.CONF', 'Custom path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+          o.on('--help', 'Displays this revoke specific help output') do |help|
+            parsed['help'] = help
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        _, non_flags, malformed_flags, unknown_flags = Utils.parse_without_raising(parser, args)
+
+        valid = valid?(results,
+                       non_flags, malformed_flags, unknown_flags,
+                       parser.help)
+
+        # if there is an exit_code then Cli will return it early, so we only
+        # return an exit_code if there's an error
+        exit_code = valid ? nil : 1
+
+        return results, exit_code
+      end
+
+      def valid?(results, non_flags, malformed_flags, unknown_flags, usage)
+        errors = []
+
+        malformed_flags.each {|f| errors << "    Missing argument to flag `#{f}`" }
+        unknown_flags.each   {|f| errors << "    Unknown flag `#{f}`" }
+        non_flags.each       {|f| errors << "    Unknown input `#{f}`" }
+
+        results['certnames'].each do |certname|
+          if CERTNAME_BLACKLIST.include?(certname)
+            errors << "    Cannot manage cert named `#{certname}` from " +
+                      "the CLI, if needed use the HTTP API directly"
+          end
+        end
+
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|e| @logger.err e }
+          @logger.err ''
+          @logger.err usage
+
+          return false
+        else
+          return true
+        end
+      end
+
+      def run(args)
+        certnames = args['certnames']
+        config = args['config']
+
+        errors = FileUtilities.validate_file_paths(config)
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|error| @logger.err "    #{error}" }
+          return 1
+        end
+
+        puppet = PuppetConfig.parse(config)
+        unless puppet.errors.empty?
+          @logger.err 'Error:'
+          puppet.errors.each {|error| @logger.err "    #{error}" }
+          return 1
+        end
+
+        revoke_certs(certnames, puppet.settings)
```

Comment:
does this need to be return revoke_certs ?

Reply:
Doesn't need to have an explicit return. Ruby will return the results of the last expression (so whatever `revoke_certs` returns).

However it might make more semantic sense for `revoke_certs` to return some kind of success/failure indicator (like a boolean or a results array) and `run` to return 0 or 1 (exit codes), rather than `revoke_certs` having to know that it should return CLI exit codes...